controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,623 @@
+}
+
+func NewSnapshotMonitor(logger logrus.FieldLogger, ds *datastore.DataStore, nodeName string, eventRecorder record.EventRecorder,
+	snapshotChangeEventQueue workqueue.Interface, syncCallback func(key string)) (*SnapshotMonitor, error) {
+
+	ctx, quit := context.WithCancel(context.Background())
+
+	m := &SnapshotMonitor{
+		baseMonitor: newBaseMonitor(ctx, quit, logger, ds, 0),
+
+		nodeName:      nodeName,
+		eventRecorder: eventRecorder,
+
+		// Periodically check snapshot disk files
+		checkScheduler: gocron.NewScheduler(time.Local),
+
+		snapshotChangeEventQueue: snapshotChangeEventQueue,
+
+		snapshotCheckTaskQueue: workqueue.NewRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 1000*time.Second),
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+		)),
+
+		inProgressSnapshotCheckTasks: map[string]struct{}{},
+
+		syncCallback:     syncCallback,
+		proxyConnCounter: util.NewAtomicCounter(),
+	}
+
+	m.checkScheduler.SingletonModeAll()
+
+	go m.Start()
+
+	return m, nil
+}
+
+func (m *SnapshotMonitor) Start() {
+	for i := 0; i < snapshotCheckWorkerMax; i++ {
+		go m.snapshotCheckWorker(i)
+	}
+
+	go m.processSnapshotChangeEvent()
+}
+
+func (m *SnapshotMonitor) processNextEvent() bool {
+	key, quit := m.snapshotChangeEventQueue.Get()
+	if quit {
+		return false
+	}
+	defer m.snapshotChangeEventQueue.Done(key)
+
+	event := key.(SnapshotChangeEvent)
+
+	m.snapshotCheckTaskQueue.Add(snapshotCheckTask{
+		volumeName:   event.VolumeName,
+		snapshotName: event.SnapshotName,
+		changeEvent:  true,
+	})
+
+	return true
+}
+
+func (m *SnapshotMonitor) processSnapshotChangeEvent() {
+	for m.processNextEvent() {
+	}
+}
+
+func (m *SnapshotMonitor) checkSnapshots() {
+	m.logger.WithField("monitor", monitorName).Info("Starting checking snapshots")
+	defer m.logger.WithField("monitor", monitorName).Infof("Finished checking snapshots")
+
+	engines, err := m.ds.ListEnginesByNodeRO(m.nodeName)
+	if err != nil {
+		m.logger.WithField("monitor", monitorName).WithError(err).Errorf("failed to list engines on node %v", m.nodeName)
+		return
+	}
+
+	defer func() {
+		m.Lock()
+		defer m.Unlock()
+		m.LastSnapshotPeriodicCheckedAt = metav1.Time{Time: time.Now().UTC()}
+	}()
+
+	for _, engine := range engines {
+		m.logger.WithField("monitor", monitorName).Infof("Populating engine %v snapshots", engine.Name)
+		m.populateEngineSnapshots(engine)
+	}
+}
+
+func (m *SnapshotMonitor) populateEngineSnapshots(engine *longhorn.Engine) {
+	snapshots := engine.Status.Snapshots
+	for _, snapshot := range snapshots {
+		// Skip volume-head because it is not a real snapshot.
+		// A system-generated snapshot is also ignored, because the prune operations the snapshots are out of
+		// sync during replica rebuilding. More investigation is in https://github.com/longhorn/longhorn/issues/4513
+		if snapshot.Name == etypes.VolumeHeadName || !snapshot.UserCreated {
```

Comment:
Need to skip removed snapshots?

Reply:
I think we can still handle removed snapshots. Any negative issue here? cc @shuo-wu

Reply:
but is it meaningful?

Reply:
I think yes. For example, the trim feature will produce a tons of removed snaphots if users enable the `UnmapMarkSnapChainRemoved`. Then, the resending the removed snapshots is still time-consuming.

Reply:
then, let's keep this first.

Reply:
BTW, a `removed` snapshot still contains in-use data before merging, so checksumming it is still necessary.